Src/Support/Google.Apis.Auth/OAuth2/AwsExternalAccountCredential.AwsSecurityCredential.cs

```diff
@@ -0,0 +1,144 @@
+                    && credentialJson != "")
+                {
+                    // Deserialize the credentials
+                    var deserializedCredentials = NewtonsoftJsonSerializer.Instance.Deserialize<AwsSecurityCredentialsResponse>(credentialJson);
+
+                    if (deserializedCredentials.IsSuccess
+                        && !string.IsNullOrEmpty(deserializedCredentials.AccessKeyId)
+                        && !string.IsNullOrEmpty(deserializedCredentials.SecretAccessKey))
+                    {
+                        return new AwsSecurityCredentials(
+                            deserializedCredentials.AccessKeyId, deserializedCredentials.SecretAccessKey,
+                            deserializedCredentials.Token == "" ? null : deserializedCredentials.Token);
+                    }
+
+                    throw new InvalidOperationException(
+                        $"The request to the metadata server security credentials endpoint {credentialEndpoint} was not succesful." +
+                        $"the response was {credentialJson}");
+                }
+
+                return null;
+
+                async Task<string> BuildSecurityCredentialsEndpointAsync()
+                {
+                    // https://github.com/aws/aws-sdk-java/blob/ccfd63c097874e3a1e9ffda7bf171769b6bd3b21/aws-java-sdk-core/src/main/java/com/amazonaws/auth/InstanceMetadataServiceCredentialsFetcher.java#L55
+                    // Fetch the list of roles attached to the instance from the credential URL as given.
+                    if (await metadataClient.FetchMetadataAsync(credentialUrl).ConfigureAwait(false) is string metadataRoleList)
+                    {
+                        // Let's use StringReader so that we don't have to worry about which end of line charater(s) to use.
+                        using var reader = new StringReader(metadataRoleList);
+                        if (await reader.ReadLineAsync().ConfigureAwait(false) is string role && role != "")
+                        {
+                            // Attach the role to the given credential URL to fetch the actual credentials.
+                            return credentialUrl.EndsWith("/") ? $"{credentialUrl}{role}" : $"{credentialUrl}/{role}";
+                        }
+                    }
+
+                    return null;
+                }
+            }
+            /// <summary>
+            /// Partial representation of a metadata server security credentials response as defined by
+            /// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials
+            /// </summary>
+            private struct AwsSecurityCredentialsResponse
+            {
+                internal const string SuccessStatusCode = "Success";
+
+                public bool IsSuccess => Code?.Equals(SuccessStatusCode, StringComparison.OrdinalIgnoreCase) == true;
```

Comment:
I suggest we add `[JsonIgnore]` to this, and `[JsonProperty]` to all the properties below to be really clear what we expect.

Reply:
And do we need to be case-insensitive for this?

Reply:
Sorry, I was unclear - I'd add JsonProperty with the actual name we expect to receive (e.g. "access_key_id" or "accessKeyId" - I don't know which it is offhand).

As well as making it clearer, it means that if we ever serialize the object (e.g. when debugging) it should be in the "right" format.

Reply:
Done. I didn't add them in the beginning because the name of the PascalCase properties matches the JSON that we recieve, but agree with that being unusual so it's best to be explicit.